### PR TITLE
feat: add pkce support

### PR DESCRIPTION
**Pull request by bshaffer**
In order to support [PKCE for OAuth2](https://datatracker.ietf.org/doc/html/rfc7636), adds the following **public** methods:

 - `OAuth2::generateCodeVerifier(): string`
 - `OAuth2::setCodeVerifier(string $codeVerifier): void`
 - `OAuth2::getCodeVerifier(): string`

**Usage**

_Authorization Request_
```php
$auth = new OAuth2(['client_id' => $clientId, 'client_secret' => $clientSecret]);

// Example of how a developer might save the code verifier in a stateless application
// NOTE: calling "generateCodeVerifier" will enable PKCE
$_SESSION['code_verifier'] = $auth->generateCodeVerifier();

// The authorization URI will include "code_challenge" and "code_challenge_method"
// querystring parameters if a code verifier is supplied
// (e.g. https://example.com?code_challenge=[base64 code_verifier]&code_challenge_method=S256)
redirect($auth->buildFullAuthorizationUri());
```
_Token Request_
```php
$auth = new OAuth2(['client_id' => $clientId, 'client_secret' => $clientSecret]);

// example of how a developer might retrieve the code verifier in a stateless application
$codeVerifier = $_SESSION['code_verifier'];

// The token URI will include "code_verifier" in the request body if a code verifier is supplied
$auth->setCodeVerifier($codeVerifier);
$token = $auth->fetchAuthToken();
```

See b/270198724

### CODE REVIEW

**Review by TimurSadykov — APPROVED** [review]
LGTM, please let Carl to look as well

**Review by clundin25 — APPROVED** [review]
Awesome stuff!

**Review by ajupazhamayil — APPROVED** [review]

File: src/OAuth2.php [hunk 1]
```diff
@@ -728,6 +747,68 @@ public function buildFullAuthorizationUri(array $config = [])
...
+     * A cryptographically random string that is used to correlate the
+     * authorization request to the token request.
+     *
+     * The code verifier for PKCE for OAuth 2.0. When set, the authorization
+     * URI will contain the Code Challenge and Code Challenge Method querystring
+     * parameters, and the token URI will contain the Code Verifier parameter.
+     *
+     * @see https://datatracker.ietf.org/doc/html/rfc7636
+     *
+     * @param string $codeVerifier
```
Comment (vishwarajanand) [inline src/OAuth2.php:768]:
```suggestion
     * @param string|null $codeVerifier
```

File: src/OAuth2.php [hunk 2]
```diff
@@ -728,6 +747,68 @@ public function buildFullAuthorizationUri(array $config = [])
         return $result;
     }
 
+    /**
+     * @return string
```
Comment (bshaffer) [inline src/OAuth2.php:751]:
```suggestion
     * @return string|null
```

**Review by vishwarajanand — APPROVED** [review]
Lgtm

**Pull request merged**